feat(mls): navigate to migrated conversation [WPB-4705]

vitorhugods opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-4705" title="WPB-4705" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-4705</a>  [Android] Observe & navigate to active 1-1 conversation
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [X] contains a reference JIRA issue number like `SQPIT-764`
    - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

When a 1:1 Proteus conversation is migrated to MLS, it actually creates a _new_ conversation, and moves all the content to the new conversation.

If that happens _while_ the user has the conversation open on the screen, a blank conversation would appear, and the user would interact with a now "defunct" conversation that should no longer be visible.

### Solutions

Observe the details of the conversation, and if it is a 1:1 conversation, check the `activeOneOnOneConversationId` with the other user and navigate to it, if it differs from the conversation being observed.

Added a new `ConversationMigrationViewModel` so it can be specialised into observing this kind of things.

It doesn't work using the current version of Kalium, as Kalium always returns `null` for `activeOneOnOneConversationId` when observing the conversation details, but I'm creating a PR in Kalium to fix it.

### Testing

#### Test Coverage

- [X] I have added automated test to this contribution

#### How to Test

You can use the DB inspector to change the `activeOneOnOneConversationId` of some user to some other conversation while you have it open.

----
#### PR Post Merge Checklist for internal contributors

- [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

marcoconti83 commented:
build again

johnxnguyen commented on wire-android-sync-engine/zmessaging/src/main/scala/com/waz/bitmap/video/VideoTranscoder.scala:
```diff
@@ -66,8 +66,7 @@ object VideoTranscoder {
 
   def apply(context: Context): VideoTranscoder =
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) new VideoTranscoder18(context)
-    else new FallbackTranscoder(context)
```
Looks like this is the only place used for `FallbackTranscoder`. We could safely delete that now.

johnxnguyen commented on wire-android-sync-engine/zmessaging/src/main/scala/com/waz/service/ContactsService.scala:
```diff
@@ -466,7 +465,7 @@ object ContactsServiceImpl {
   }
 
-  lazy val Visible = if (SDK_INT >= LOLLIPOP) Some(s"${Col.Visible} = 1 OR ${Col.InDefaultDirectory} = 1") else None
+  lazy val Visible = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) Some(s"${Col.Visible} = 1 OR ${Col.InDefaultDirectory} = 1") else None
```
Do we need this conditional? The all builds should now be >= `LOLLIPOP`.

marcoconti83 replied:
Correct!

johnxnguyen commented on wire-android-sync-engine/zmessaging/src/main/scala/com/waz/service/NetworkModeService.scala:
```diff
@@ -73,7 +74,7 @@ class DefaultNetworkModeService(context: Context, lifeCycle: UiLifeCycle) extend
   //this method doesn't really belong here, but it's only used for tracking - just tells us if the device was in doze mode or not
   //fixme get the version code from android sdk
-  @TargetApi(23)
+  @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
```
`JELLY_BEAN_MR2` is API level 18. I think you mean `M` here.

marcoconti83 replied:
Well spotted :)

johnxnguyen commented on wire-android-sync-engine/zmessaging/src/main/scala/com/waz/utils/Locales.scala:
```diff
@@ -44,7 +43,7 @@ object Locales extends DerivedLogTag {
     } yield locale) getOrElse Locale.getDefault
 
-  lazy val bcp47 = if (SDK_INT >= LOLLIPOP) AndroidLanguageTags.create else FallbackLanguageTags.create
+  lazy val bcp47 = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) AndroidLanguageTags.create else FallbackLanguageTags.create
```
Again, I don't think you need this conditional, which also means you can delete `FallbackLanguageTags`

vitorhugods committed:
feat(mls): redirect users to migrated
one-on-one conversation

This commit introduces a new ViewModel `ConversationMigrationViewModel.kt` for tracking conversation migrations. It observes details of a conversation and checks if this conversation was migrated to a different one-on-one conversation. If it was, it updates the target conversation with the ID of the active one-on-one conversation.

typfel committed:
chore: update kalium ref

vitorhugods committed:
feat(mls): navigate to migrated conversation

AndroidBob commented:
Build [1245](https://10.10.124.134/job/AR-build-pipeline/1245/) **failed**.

AndroidBob commented:
Build [1246](https://10.10.124.134/job/AR-build-pipeline/1246/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.5.0(7807301)(7807301)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.5.0(7807301)(7807301)-staging-compat.apk)

yamilmedina reviewed (approved):
Looking awesome 🧉

AndroidBob commented:
Build [1253](https://10.10.124.134/job/AR-build-pipeline/1253/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.5.0(7813002)(7813002)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.5.0(7813002)(7813002)-staging-compat.apk)
